### DISCUSSION TITLE
How to do Geospatial Analysis with Python

### DISCUSSION THREAD

**Discussion by Kirs10-Riley** [kbuzard/labs#41] [General]
# How to do Geospatial Analysis with Python

information gathered from: https://kodu.ut.ee/~kmoch/geopython2019/L1/Geometric-Objects.html#overview-of-geometric-objects-and-shapely-module

## [The basics to Spatial data modeling](https://kodu.ut.ee/~kmoch/geopython2019/L1/Geometric-Objects.html#overview-of-geometric-objects-and-shapely-module) 

-   **Point** - object represents a single point in space. Points can be either two-dimensional (x,y) or three-dimentional (x,y,z).

    How to create a point:

    -   `Point_name = Point(X, Y)`

    -   `print(Point_name)` - to see the point

    -   3D coordinates will have a capital Z-letter in front of the coordinates

    Point attributes and functions:

    -   extracting coordinates

        -   `point_name.xy` - retrieves x and y coordinates
        -   `point_name.x` - retrieves x
        -   `point_name.y` - retrieves y

    -   Distance between points

        -   `point_name1.distance(point_name2)`

        -   side note: this is linear distance, to calculate real distance on Earth's surface more steps are needed and you must import "math" and "pyproj"

-   **Polygon** - object representing a filled area that consists of a list of at least three coordinate tuples that forms the outer ring and a (possible) list of hole polygons.

    How to create a line:

    -   `polygon_name = Polygon([(x,y), (x,y), (x,y), ...])`

    -   `Polygon(shell= *exterior polygon*, holes= *interior polygon*)` - polygon with a hole

        Polygon attributes and functions:

    -   polygon_name.centroid

    -   polygon_name.area

    -   polygon_name.bounds

    -   polygon_name.exterior

    -   polygon_name.length

**Reading X/Y Coordinates from Text files**

pandas is a popular python-based analysis toolkit which can be used to parse multiple file formats. It can be used to read a file.

first you would

`import pandas as pd`

then find the path to your working file

`df_name = pd.read_csv(path to your working file)`

this will bring the data set into your work space

**functions to manipulate the data**

-   `pd.set_option('max_columns', \#)` - this will only print a certain amount of collumns

-   `print(df_name.head(\#)` - this will print the first \# of rows of the data set

## [Intro to Geopandas](https://kodu.ut.ee/~kmoch/geopython2019/L2/geopandas-basics.html)

GeoPandas extends the datatypes used by pandas to allow spatial operations on geometric types. It allows python to easily read downloaded spatial data using the `gpd.from_file()` function

**To use**

`import geopandas as gpd`

add the filepath to your workspace 

`data_name = gpd.from_file(*insert your filepath relative to your working directory*)`

to see data on a map use the `.plot()` -function

`data_name.plot()`

**From here the data can be read/ analyzed using commands such as: **

`data_name.area` area of the data frame 
`data_name.area.max()` max area
`data_name.area.mean()` mean area

## [Coordinate Reference Systems (CRS)](https://kodu.ut.ee/~kmoch/geopython2019/L2/crs-projections.html)

A CRS tells python how coordinates related to places on Earth. This is important because the geometric shapes in a GeoDataFrame are simply a collection of coordinates in an arbitrary space. 

-> CRS is often referred to as SRS (Spatial Reference System)

**There are two common types:**
1. Geographic Coordinate Systems 
2. Projected Coordinate Systems 

**Geographic coordinate system (GCS)**

Uses ellipsoidal surface to define locations on the Earth and has three parts: 
- *A datum* -  an ellipsoidal (spheroid) model of the Earth to use. Common datums include WGS84 (used in GPS).
- A prime meridian
- Angular unit of measure

**Projected coordinate systems (PCS)** 

Defines a flat 2D cartesian surface. Unlike the GCS, a projected coordinate system has constant length, angles, and areas across the two dimensions. A projected coordinate system is always based on a geographic  coordinate system that references a specific datum.

consist of: 
- GCS 
- projection method
- projection parameters (standard points and lines, Latitude of Origin, Longitude of Origin, etc.)
- Linear units 

**Defining and changing CRS in Geopandas**

GeoDataFrames that is read from a shape file should always contain information about the coordinate system in which the data is projected. 

To see the current coordinate reference system from a data frame use the command `.crs` 

to learn more about the spacial reference system use these links: 
- [www.spatialreference.org](http://spatialreference.org/)
- [www.epsg.io](https://epsg.io/)
- [www.proj4.org](https://proj.org/operations/projections/index.html)
- [www.mapref.org](http://mapref.org/CollectionofCRSinEurope.html)

to change the CRS use the `.to_crs()` -function 

ex: `df_name.to_crs(epsg = xxxx)`

**Making a Map**

using the df we can use these command to create a map

`import matplotlib.pyplot as plt`
`data_name.plot(facecolor= 'color name')` - prints a map into the desired color
`plt.title("title")` - adds a title 
`plt.tight_layout()` - removes the white space around the plot

**Comment by kbuzard** [comment]
Thanks @Kirs10-Riley ! You found a great resource and did a fantastic job summarizing it. Here are a few ways to optimize your explainer:
1. You can take out the part about LineStrings. I've never had to use these.
2. Some of the bold formatting didn't get applied -- I think there may be extra spaces that somehow interfere with Markdown understanding what you want.
3. Could you add links for each page where the info came from? For instance, if all of the info on CRS comes from one page, add a link to the title with [Title](url)
